python/dlisio/ext/lis.cpp

```diff
@@ -120,73 +156,252 @@ noexcept (false) {
+noexcept (false) {
+    using namespace std;
+
+    double x;
+    const char* indexfmt = fconf.indexfmt.c_str();
+    switch (*indexfmt) {
+        case LIS_FMT_I8:     { x = from_buf< int8_t  >(src, end, fconf); break; }
+        case LIS_FMT_I16:    { x = from_buf< int16_t >(src, end, fconf); break; }
+        case LIS_FMT_I32:    { x = from_buf< int32_t >(src, end, fconf); break; }
+        case LIS_FMT_F16:    { x = from_buf< float   >(src, end, fconf); break; }
+        case LIS_FMT_F32:    { x = from_buf< float   >(src, end, fconf); break; }
+        case LIS_FMT_F32LOW: { x = from_buf< float   >(src, end, fconf); break; }
+        case LIS_FMT_F32FIX: { x = from_buf< float   >(src, end, fconf); break; }
+        case LIS_FMT_BYTE:   { x = from_buf< uint8_t >(src, end, fconf); break; }
+        case LIS_FMT_STRING:
+        case LIS_FMT_MASK: {
+            const auto msg = "Unsupported datatype for index channel, was: ";
+            throw std::runtime_error(msg + fconf.indexfmt);
+        }
+        default: {
+            const auto msg = "Invalid datatype for index channel, was: ";
+            throw std::runtime_error(msg + fconf.indexfmt);
         }
     }
+    channel.update(x);
+    return src;
 }
 
-void read_data_record( const std::string& fmt,
-                       const lis::record& src,
+void read_frame( const char*&       ptr,
+                 const char*        end,
+                 unsigned char*&    dst,
+                 indexchannel&      index,
+                 const frameconfig& fconf )
+noexcept (false) {
+    bool fast = ( fconf.samples > 1 ) ? true : false;
+
+    /* Writing the index to dst
+     *
+     * The trivial case is when the index channel is recorded as a regular
+     * channel (mode=0) and we don't have fast channels. Then, there is no need
+     * for the intermediate storage of the index values and we can write
+     * directly from record (ptr) to dst.
+     */
+    if ( fconf.mode == 0 and not fast ) {
+        const char* indexfmt = fconf.indexfmt.c_str();
+        read_sample( indexfmt, ptr, end, dst );
+    } else {
+        dst = write_index( dst, fconf, index.index() );
+    }
+
+    /* Read the curves
+     *
+     * The index written to dst in its entirety at this point (Including
+     * potential subframes). Both fmt, ptr and dst are appropriately aligned
+     * such that we can go ahead and read all the non-index channels.
+     */
+    const char* fmt = fconf.fmtstr.c_str();
+    while ( *fmt != LIS_FMT_EOL ) {
+        /* Suppression of Channels
+         *
+         * Suppression doesn't care about the structure of the recorded
+         * channel. I.e.  number of samples or number of entries. It simply
+         * skips over all bytes in the record (ptr) that is associated with a
+         * channel.
+         */
+        if ( *fmt == LIS_FMT_SUPPRESS) {
+            char* next;
+            ptr += std::strtol(++fmt, &next, 10);
+            fmt = next;
+            continue;
```

Comment:
Ops, I meant to remove I in read_sample. Because the fmtstr for suppress does not apply to a single sample, it applies to the entire channel. Hence it makes more sense to the check _here_